src/libcore/mem.rs

```diff
@@ -1092,37 +985,108 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
 /// // Create an explicitly uninitialized reference. The compiler knows that data inside
 /// // a `MaybeUninit<T>` may be invalid, and hence this is not UB:
 /// let mut x = MaybeUninit::<&i32>::uninit();
 /// // Set it to a valid value.
-/// x.write(&0);
+/// unsafe { x.as_mut_ptr().write(&0); }
 /// // Extract the initialized data -- this is only allowed *after* properly
 /// // initializing `x`!
 /// let x = unsafe { x.assume_init() };
 /// ```
 ///
 /// The compiler then knows to not make any incorrect assumptions or optimizations on this code.
-//
-// FIXME before stabilizing, explain how to initialize a struct field-by-field.
+///
+/// ## Initializing an array element-by-element
+///
+/// `MaybeUninit<T>` can be used to initialize a large array element-by-element:
+///
+/// ```
+/// use std::mem::{self, MaybeUninit};
+/// use std::ptr;
+///
+/// let data = unsafe {
+///     // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+///     // safe because we type we are claiming to have initialized here is a
+///     // bunch of `MaybeUninit`, which do not require initialization.
+///     let mut data: [MaybeUninit<Vec<u32>>; 1000] = MaybeUninit::uninit().assume_init();
+///
+///     // Dropping a `MaybeUninit` does nothing, so if there is a panic during this loop,
+///     // we have a memory leak, but there is no memory safety issue.
+///     for elem in &mut data[..] {
+///         ptr::write(elem.as_mut_ptr(), vec![42]);
+///     }
+///
+///     // Everything is initialized. Transmute the array to the
+///     // initialized type.
+///     mem::transmute::<_, [Vec<u32>; 1000]>(data)
+/// };
+///
+/// println!("{:?}", &data[0]);
+/// ```
+///
+/// You can also work with partially initialized arrays, which could
+/// be found in low-level datastructures.
+///
+/// ```
+/// use std::mem::MaybeUninit;
+/// use std::ptr;
+///
+/// unsafe {
+///     // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+///     // safe because we type we are claiming to have initialized here is a
+///     // bunch of `MaybeUninit`, which do not require initialization.
+///     let mut data: [MaybeUninit<String>; 500] = MaybeUninit::uninit().assume_init();
+///     // Count the number of elements we have assigned.
+///     let mut data_len: usize = 0;
+///
+///     for elem in &mut data[0..500] {
+///         ptr::write(elem.as_mut_ptr(), String::from("hello"));
+///         data_len += 1;
+///     }
+///
+///     // For each item in the array, drop if we allocated it.
+///     for elem in &mut data[0..data_len] {
+///         ptr::drop_in_place(elem.as_mut_ptr());
+///     }
+/// }
+/// ```
+///
+/// ## Initializing a struct field-by-field
+///
+/// There is unfortunately currently no supported way to create a raw pointer or reference
+/// to a field of a struct inside `MaybeUninit<Struct>`. That means it is not possible
+/// to create a struct by calling `MaybeUninit::uninit::<Struct>()` and then writing
+/// to its fields.
```

Comment:
But I guess I can get around this in the special case that the struct is a newtype with `repr(transparent)` or (`repr(C)`), by calling `MaybeUninit::uninit::<T>()` where `T` is the wrapped type? Not sure if it's worth noting.

Reply:
Okay, scratch that. Nevermind. If there's just one member, then it does not matter anyway and I don't need to use `T`, I can just use the wrapper.

Reply:
The problem is basically that we don't have https://github.com/rust-lang/rfcs/pull/2582. I wonder if I should put a link to the RFC into the docs? I haven't seen that elsewhere though.

Reply:
> But I guess I can get around this in the special case that the struct is a newtype with `repr(transparent)` or (`repr(C)`), by calling `MaybeUninit::uninit::<T>()` where `T` is the wrapped type? Not sure if it's worth noting.

Note that in the `#[repr(C)]` case since the order, alignment, and padding of fields is known you could manually calculate the offsets and apply them to the base pointer. A pain for sure, but totally doable (I'm doing that in a project of mine).